source/slang/hlsl.meta.slang

```diff
@@ -9335,7 +9378,7 @@ for(const char* T : {"int64_t", "uint64_t"})
 {
 }}}}
 [ForceInline]
-[require(cuda_glsl_hlsl_spirv, atomic_glsl_hlsl_cuda)]
+[require(cuda_glsl_hlsl_spirv, atomic_glsl_hlsl_cuda_metal)]
```

Comment:
It should not be a capability error.

The reason why is because of how the capability system resolves capabilities:
`atomic_glsl_hlsl_cuda_metal` has the targets `cuda` `glsl` `hlsl` `metal` `spirv`
`cuda_glsl_hlsl_spirv` has the target `cuda` `glsl` `hlsl` `spirv`

`cuda_glsl_hlsl_spirv + cuda_glsl_hlsl_metal_spirv` is equivalent to `cuda_glsl_hlsl_spirv`. As a result, the `metal` target is not a requirement of the function

Reply:
Thank you for the explanation. But I am still having a trouble understanding it.
Would you mind explain little more of how `+` sign works? I thought [require(xx,yy)] means "xx && yy" where `&&` means both conditions should be true. I was reading it as "the function requires the target to be (cuda || glsl || hlsl || spirv) && (atomic_cuda || atomic_glsl || atomic_hlsl || atomic_metal || atomic_spirv)", but I guess I am reading it wrong.

Reply:
'+' is similar to `&&`. What has to be remembered when evaluating is that we discard invalid sets. Note that, valid sets must only have 1 key (abstract-atom-of-type) fill a keyhole (abstract-atom-type) per set.

What this means is `(cuda || glsl || hlsl || spirv) && (atomic_cuda || atomic_glsl || atomic_hlsl || atomic_metal || atomic_spirv)` is a correct representation.
The way we evaluate this expression can be thought of in the following way:
1. `(cuda || glsl || hlsl || spirv) && (atomic_cuda || atomic_glsl || atomic_hlsl || atomic_metal || atomic_spirv)` is equivalent to the expression of `(cuda && atomic_cuda || cuda + && atomic_glsl || cuda + && atomic_hlsl || cuda + && atomic_metal || cuda + && atomic_spirv || glsl + atomic_cuda ||  glsl + && atomic_glsl || glsl + && atomic_hlsl || ...)
2. Since invalid sets are discarded, we reflect this when simplifying the expression: `(cuda + atomic_cuda || glsl + atomic_glsl || hlsl + atomic_hlsl || spirv + atomic_spirv)`
3. It is important to notice that `cuda + atomic_metal`, `glsl + atomic_metal`, `hlsl + atomic_metal`, `spirv + atomic_metal` are all invalid, so we discard these sets.